### PR TITLE
GUI for Changing Settings

### PR DESCRIPTION
**Pull request by thelegendofbrian**
# In this update
- Cross-OS compatible GUI to change most config options without reloading the client
	- Right click the system tray icon or press Ctrl + O to get to the settings GUI
	- Use either your system's notification system or a custom one
- A slew of new features to fill in the GUI:
	- Default client size
	- Popup notifications for:
		- PMs
		- Duel requests
		- Logout warning
		- Low HP
		- High fatigue
- Configurable keybinds for most boolean settings (addresses #22)

# Notes
- Since this is such a large change, it might be wise to put this in a new branch while it gets tested on other OS's since there are a lot of OS-specific changes.
- There are a lot of things to do to finish off what @Talkarcabbage and I started. There are tons of TODOs everywhere for improvements that would be nice.

# What it looks like
**The settings GUI (updated)**
ui3

**Triggering a notification in Windows 10 with the system notification system**
tray-icon-system

**Triggering a notification in Windows 10 with the custom notification system**
tray-icon-custom

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,7 +15,7 @@
 *I will not accept any game automation features (macroing, etc.)*
 
 ## Features
-- Game resolution changing (about 80% complete)
+- Game resolution changing
```
Comment (OrN) [inline README.md]: Why was this removed? The 80% complete was talking about various interfaces not centered/resized to fit resizable, but I guess it's not important

File: src/Client/ConfigWindow.java [hunk 2]
```diff
@@ -0,0 +1,1249 @@
+package Client;
```
Comment (OrN) [inline src/Client/ConfigWindow.java]: GPL license header is missing

File: src/Client/KeybindSet.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package Client;
```
Comment (OrN) [inline src/Client/KeybindSet.java]: GPL license header is missing

File: src/Client/NotificationsHandler.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
+package Client;
```
Comment (OrN) [inline src/Client/NotificationsHandler.java]: GPL license header is missing

File: src/Client/TrayHandler.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
+package Client;
```
Comment (OrN) [inline src/Client/TrayHandler.java]: GPL license header is missing

**Comment by OrN** [comment]
The tray icon is pretty nice for opening the settings GUI. It keeps the client from looking like most bot clients/osbuddy, etc.

File: README.md (hunk 1 above)
Comment (thelegendofbrian) [inline README.md]: Oh, I thought it was implemented. I'll change it back.

**Comment by thelegendofbrian** [comment]
This commit should address all the issues mentioned so far.

**Comment by OrN** [comment]
The settings window could use the icon from assets

**Comment by Hubcapp** [comment]
You'd want libnotify for linux [(example code)](https://wiki.archlinux.org/index.php/Desktop_notifications), and growl for macOS [(possibly using this library)](https://github.com/jitsi/growl4j). If you don't want to do it, maybe I'd do it later. For now, the fallback notification system looks great.

Also will need to update "::help keys" to read keybinding changes, although its use is questionable when you could just use the gui to check them and should probably just be removed (can leave text command ::help though). Maybe ::help and the first-time-login message could be reworded to give instructions on opening the settings panel.

btw, tested on Debian

[ INFO] Loading resource: /assets/icon.png
[ INFO] Loading resource: /assets/icon-small.gif
[ INFO] Loading resource: /assets/icon.png
[ INFO] Loading resource: /assets/notification_close.png
[ INFO] Loading resource: /assets/notification_close_highlighted.png
[ INFO] Loading resource: /assets/notification_background.png
[ INFO] Loading resource as stream: /assets/OpenSans-Regular.ttf
[ INFO] Loading resource as stream: /assets/OpenSans-Bold.ttf
[ INFO] Loading resource as stream: /assets/notification.wav
Exception in thread "main" java.lang.IllegalArgumentException: Invalid format
	at org.classpath.icedtea.pulseaudio.PulseAudioDataLine.createStream(PulseAudioDataLine.java:142)
	at org.classpath.icedtea.pulseaudio.PulseAudioDataLine.open(PulseAudioDataLine.java:99)
	at org.classpath.icedtea.pulseaudio.PulseAudioDataLine.open(PulseAudioDataLine.java:283)
	at org.classpath.icedtea.pulseaudio.PulseAudioClip.open(PulseAudioClip.java:402)
	at org.classpath.icedtea.pulseaudio.PulseAudioClip.open(PulseAudioClip.java:453)
	at Client.NotificationsHandler.loadNotificationSound(Unknown Source)
	at Client.NotificationsHandler.initialize(Unknown Source)
	at Client.Launcher.main(Unknown Source)

may look into this for you, see if I can fix it

**Comment by Hubcapp** [comment]
added a catch for that new exception where it told me to, got this new error with game sound enabled:

Error: client.LD(87,34,120) client.SB(-26345) client.RD(0) client.MA(119) | org.classpath.icedtea.pulseaudio.PulseAudioClip.setMicrosecondPosition:503 Client.NotificationsHandler.playNotificationSound Client.NotificationsHandler.displayNotification Game.Client.messageHook client.a client.a client.K client.J client.e e.run java.lang.Thread.run:745 | java.lang.IllegalStateException: Line not open
error_game_crash

I guess it doesn't expect to not have the line open when trying to play a sound.

Once I disabled audio in settings, I was able to see a notification, but because of my strange monitor configuration, it was displayed partially offscreen and btw doesn't look as nice as the picture above. 

Also, tray icon has gray background which doesn't look very nice. Here's a screenshot: https://i.sli.mg/tGKk5V.png

**Comment by thelegendofbrian** [comment]
I was aware of the tray icon not supporting transparency in Linux. Java has known about it for 10 years and hasn't fixed it yet: http://bugs.java.com/bugdatabase/view_bug.do?bug_id=6453521 A workaround would be to just make the icon a box without transparency. If someone wants to make another tray icon, they can, I just kinda slapped something together for the project and made sure the name was visible on different background colors.

**Comment by thelegendofbrian** [comment]
The notification that you had appear off-screen is created by your desktop environment is my understanding. I call that notification the same way I call the Windows built-in notification system. I tested system notifications on Ubuntu GNOME and it did not render off-screen so it's probably due to your monitor configuration like you said.

**Comment by thelegendofbrian** [comment]
In regards to the `::help` commands, there still a niche use for them even with the GUI. Chat commands that type things for you should remain in the help pages (i.e. `::cmb`, `::cmbnocolor`, `::total`, `::fatigue`, `::next_[skillname]`, `::[skillname]`)

**Comment by Hubcapp** [comment]
the GUI for configuration options is great. sorry if my bug report came off negative. :-) only thing I would change about it (functionally) is to apply changes immediately instead of requiring user to click the apply button. Would make it snappier I think and remove the problem of the user wondering if they remembered to apply changes or not (could also be achieved by graying out apply button if no changes to apply)

I agree that ::help text [x] should continue to exist, but yeah, either ::help keys [x] needs to be removed or updated to reflect keybinding changes. Whichever seems more appropriate, i don't have any strong feelings about it.

Looking at my other java application tray icons, Jitsi just uses a cyan solid background behind their icon and JDownloader has solid white corners. I'm not sure how the tray icon for RSC+ could be improved beyond something like putting the RSC grass gradient behind it or something (don't like solid color boxes), but I wouldn't want that for Windows or OS X b/c it probably will not look as good as a transparent background. The icon you drew up looks great with transparency. Could detect OS (might already do that? don't remember) and switch tray icon based on that.

As for rendering off screen, I understand that my configuration is esoteric, and given the low volume of active RSC players (about 8 to 30 players online at any given time once bots are removed) it is probable that I would be the only user to have that issue. :-) It'd be fine to ignore (especially once native notifications are implemented there's almost no point fixing it), but I'll fix it for my window manager if I get the time. It's AwesomeWM, by the way, which is tiling.

**Comment by thelegendofbrian** [comment]
Oh you're fine, I didn't perceive it as negative. I'll look into making the tray icon and notifications OS-specific as time permits.

**Comment by Hubcapp** [comment]
Just want to make sure that you get some positive feedback here. When you work really hard on something and you're proud of it, and then you just get "lol it crashed and doesn't look as good as I thought it might b/c of my 1/10000000 1337 configuration" as feedback that sucks. You've made some really good changes here. It really makes me happy to see RSC get the attention and love this project has given it.

**Comment by Talkarcabbage** [comment]
Wasn't perceived as negative, but thanks for saying so anyway :)
As for the notification appearing off-screen, does that happen with the built-in notifications or just with system notifications enabled? I'm not sure there's a way to adjust the location of system notifications, though the built-in GUI's location can be tweaked as needed.
OS detection should be do-able for the icon, in theory, though the specifics might be more complicated.
Will work on the audio crash situation, it's _probably_ resolvable easily.

**Comment by Hubcapp** [comment]
Okay, I tested it with the box for "use system notifications if available" unchecked and got this:

https://i.sli.mg/ysjSeb.png

The screenshot from before is with "use system notifications if available" toggled "On".

**Comment by Hubcapp** [comment]
here's another without "use system notifications"

https://i.sli.mg/7xRDRq.png

I don't know why it's different. Will test your new linux audio commit.

**Comment by Hubcapp** [comment]
No crash anymore, also no notification audio. (checked pavumeter and pavucontrol to confirm) Normal game audio works.

**Comment by Talkarcabbage** [comment]
Was the audio notification still not working prior to that as well as after?

**Comment by Hubcapp** [comment]
Before the entire client crashed while loading the resource, and I was only able to get around it by adding an extra catch statement. The catch just "ignored" that the resource couldn't be loaded, so it was not working before or after that latest commit, but this commit is an improvement because it no longer crashes.

**Comment by Hubcapp** [comment]
I was able to reproduce the second behavior in https://i.sli.mg/7xRDRq.png by changing AwesomeWM layout from floating to tile, and then back to floating. I also needed to have an extra window open but minimized on the same screen as the notification. Really not sure you'd want to fix it, but the non-native popup is I'd say "not useful", or at least "broken enough that I wouldn't use it" on a tiling window manager. Also, the notification appears on whatever monitor the cursor is on. It only appears top monitor in my screenshots because that's where it always was when I was sending messages to myself.

**Comment by Talkarcabbage** [comment]
The audio should work properly now on linux. It seems it was being stubborn and buggy while working on fine on Windows. I blame linux, in the option of blaming linux or blaming java.
Also the notifs window should look less derp on linux, as it will no longer have the shadows.
I'm not sure what to do about the tiling window manager, other than to suggest alternative window managers that don't modify the size of size-preset-windows.

**Comment by OrN** [comment]
Is this ready to be merged?

**Comment by Hubcapp** [comment]
Maybe. I tested latest changes on my tilingwm special edgecase linux distro and while the notifications aren't perfect, it works fine to turn them off if nothing else. Don't think I would have to at this point, System notification (while broken in a few ways, listed above mostly still applicable) works well enough now that I'd keep it enabled just for the audio cues. Audio cue on PM is ___really___ cool. Only issue I have is that when using system notification on linux, the bubble never autodismisses, and it'll push my windows around until I close it. A more native system notification using libnotify would be ideal, but not worth postponing merge imo. Non-system notification still very broken, which kind of defeats point of having a "non native" notification at all imo, although I guess it probably works fine everywhere that isn't special edgecase tiling wm land.

Might be nice to toggle audio cues separate of the visual system notification. Should be pretty easy, and would let me (and possibly other users) use audio cue without broken system notifications in edgecase land.

Also got around to testing on OSX and it works fine there too. It's a bit silly though, native system notification checked, same issue as on linux with this box that never autodismisses. Not sure if this is new, but I can't resize above a certain resolution on OSX either (never tried client there). Also can't change camera angle sometimes... see https://i.sli.mg/CyD46j.png

Settings menu is totally rad and should be merged into main asap.

**Comment by Hubcapp** [comment]
I guess 

* immediately apply settings without needing Apply/cancel button

would be nice too. Only time I would want to cancel settings is if I messed up all my keybinds.

**Comment by thelegendofbrian** [comment]
We'll try to knock out some of these today, however neither @Talkarcabbage nor I have access to a machine running OS X. So if someone with OS X could check to see if our changes broke it or it was already like that, that would be appreciated.

**Comment by Hubcapp** [comment]
Okay, I checked current release, and the resizing on OSX is an issue there too. So you're clear. ;)

Attached is an image of the maximum resolution in both x and y direction before it stops rendering. It seems to be about 1225x831 which doesn't mean anything to me. (was hoping it'd be something like 1280x1024)

https://i.sli.mg/q19YtJ.png

**Comment by Talkarcabbage** [comment]
Linux and java GUIs are the most frustratingly broken combination I have ever laid eyes on. Completely depending on your window manager, the following unpredictable/predictable behaviors _may_ occur:
* The window may cause half the screen to not render (likely a result of the window manager setting the size of the jframe, which it should not as it's being set manually. Attributing this to your specific monitor and window manager setup, as this has been a non-issue in tests I've done)
* The window may have an icon in the "taskbar" of some window managers (which it shouldn't, this is an OS problem; setting the frame's type as is done in code should remove this icon.
* The icon may sometimes be a penguin on openJDK. No appreciable explanation for this.
* The custom notification area may for some reason have extra (black, or possibly another color) border, but this is mostly resolved with a recent push, in theory. It was tested limitedly (notable on several distros, not sure which specific desktops cause it. Does not occur on Debian)
* The notification window may not repaint correctly, and the icon may be completely transparent while the close button is invisible (Notable on Debian desktop)
* The font looks ugly. Apparently this is a common problem solved by installing some third party packages that I don't know the availability of.

Not sure how to fix any of these graphical issues. None of them appear on Windows, and they are unpredictable across different window managers on linux. Likely a result of some transparency/opaqueness settings that should work that do not.

Not entirely sure what the resizing bug is, but if it already exists in the current release code our changes (i.e. the on-start-client-size option) shouldn't break it and would in theory also work with a fix if it is implemented.

**Comment by Hubcapp** [comment]
>     The window may cause half the screen to not render (likely a result of the window manager setting the size of the jframe, which it should not as it's being set manually. Attributing this to your specific monitor and window manager setup, as this has been a non-issue in tests I've done)

I'd like to take the blame for this as the user, yeah. There's a config option that I have changed specifically to ignore what size the window wants to be (size_hints_honor = false). This is generally more useful than letting the window set its own size; the way you're using a window here is also an edge case. ;) Other tiling WMs might not have that option, don't know.

>     The window may have an icon in the "taskbar" of some window managers (which it shouldn't, this is an OS problem; setting the frame's type as is done in code should remove this icon.

I've seen jdownloader manage to spawn a window in linux without this icon, it could be possible to remove with a different implementation, but don't bother.

>     The icon may sometimes be a penguin on openJDK. No appreciable explanation for this.

Probably because the mascot for java is copyright by Oracle and openJDK aims to be a [Free software](https://www.gnu.org/philosophy/free-sw.en.html) alternative to the official JDK.

>     The custom notification area may for some reason have extra (black, or possibly another color) border, but this is mostly resolved with a recent push, in theory. It was tested limitedly (notable on several distros, not sure which specific desktops cause it. Does not occur on Debian)

I'm using Xfce on Debian as the base, then installed awesomewm over it. Seemed to happen on OSX, (no border on top or left, only on right and bottom) but didn't mention it.

>     The notification window may not repaint correctly, and the icon may be completely transparent while the close button is invisible (Notable on Debian desktop)

I don't have close buttons so I didn't notice this issue.

>     The font looks ugly. Apparently this is a common problem solved by installing some third party packages that I don't know the availability of.

Yeah, font rendering in java is ugly. That's up to the user to put up with or try to fix. (see https://wiki.archlinux.org/index.php/Java_Runtime_Environment_fonts)

**Comment by Talkarcabbage** [comment]
Oh, uh. The latest version should look a lot better, as of my last push. Tested with debian, ubuntu, mint desktops and each looked fine. The fonts are still an issue though, but yeah, that's not really within scope of the program. Let me know if the window still looks wonky.

**Comment by Hubcapp** [comment]
>Oh, and with that change the "whole half of your screen not updating" may turn into "whole half of your screen taken up by a weird looking jframe window" on that tiling manager.

Yep, you correctly predicted that behavior lol. It's a lot better, yeah. it has correctly drawn bubble, then gray background for area that isn't meant to be seen.

non-native notification on OSX is great now (autodismisses, looks good, correct position), native is "meh", same as in picture before but correctly positioned now. still doesn't autodismiss (on osx only), but does on linux.

**Comment by Talkarcabbage** [comment]
It doesn't auto-dismiss on OSX? That's odd. Is that a consistently doesn't dismiss, or a sometimes doesn't? Any logged console errors?

**Comment by Hubcapp** [comment]
Nothing in the console, no. For native notifications on OSX It just spawns the notification window and waits for OK to be pressed still. Non-native autodismisses.

**Comment by Hubcapp** [comment]
Blaming the OS is like building a website that only works in IE6 and then blaming Netscape for being different. Even though IE6 had 95% market share, it doesn't mean everything they did was great and everyone else is stupid and broken. It's just different requirements for different environments (of which there are a myriad of choices outside of Windows), and yeah handling them all can be frustrating. Sorry, don't mean to be offensive, just got put off by the way you're thinking of non-windows operating systems.

Notifications are great now everywhere, I think ready to be merged. This patch makes RSC+ a lot better. :-)

As feature requests, these are things that could be completed before/after merge either by me/you guys. I won't work on them until after merge though

* Would like to only enable audio notification only, without bubble notification
* Get rid of cancel/apply/ok button probably everywhere except keybinds tab, but maybe even there. Can just instantly apply instead.
* Fix non-native notification for tiling wm by checking out what jdownloader does and learning from them
* Implement libnotify & growl as "real" system native notifications
* Add slider for how long xp drop bar should be displayed (maximum notch should be "forever")

**Comment by Talkarcabbage** [comment]
I mean, the reason the non-native notification system was added at all is if the OS in question has a bad or nonexistent notification system. There's nothing wrong with attributing unexpected and unintended product behavior to a quirk of the operating system, as it's entirely dependent on which OS you're using and what you have installed on it as to how the notifications behave and are rendered in the case of the native ones (and is also why growl and libnotify are on the table as alternatives). I didn't say any of the above operating systems are inferior or bad choices and not worth coding for, I'm just saying that those OS's have weird behaviors that make their built-in native solutions sub-optimal. Again, that's originally why the non-native notifications were implemented (they just ended up buggy and wonky looking for a while).

* Audio notification do-able, depends on implementation though
* Instant save probably do-able, would need some testing
* Not sure what JDownloader does. This may be a post-merge project.
* Growl and libnotify seemed a tad complicated, though @thelegendofbrian was the one looking at those. May be a post-merge project.
* Not sure how this/if this is actually implemented in code already. If it's not, for that reason it may be something to do post-merge.

**Comment by thelegendofbrian** [comment]
Don't merge quite yet, still need to push my OS-dependent tray icons, but still testing.

**Comment by Hubcapp** [comment]
Regarding a potential slider for xp drop bar display length, I made a variable for it but it's not exposed as editable in config yet. it's a TODO.

**Comment by Hubcapp** [comment]
I got this crash standing in Lumbridge cow field talking to a friend.

Error: ua.RA({...},0,256,0) client.AB(13) client.JD(false) | java.util.TimSort.mergeLo:747 java.util.TimSort.mergeAt:483 java.util.TimSort.mergeCollapse:410 java.util.TimSort.sort:214 java.util.TimSort.sort:173 java.util.Arrays.sort:659 java.util.Collections.sort:217 Game.Renderer.present ua.a client.f client.b e.run java.lang.Thread.run:745 | java.lang.IllegalArgumentException: Comparison method violates its general contract!
error_game_crash

**Comment by Talkarcabbage** [comment]
So the bug exists in ORN's master then?

**Comment by Hubcapp** [comment]
Yup @Talkarcabbage

**Comment by Talkarcabbage** [comment]
If this change doesn't fix your window manager making the notifications take up half the screen, then it's because you have your manager set to ignore size hints. At your leisure, see if this change makes the non-native notifications not resize beyond their intended size.

**Comment by thelegendofbrian** [comment]
In regards to making the tray icon change to non-transparent and properly sized on Linux, we did some testing on a few Linux distros and it appears that the size of the tray icon is sometimes 22x24 and sometimes 16x16, however 22x24 seems more popular. Without making special exceptions for individual distros and desktop environments, we can't get rid of the grey background on the tray icon since we'll need to keep the size 16x16 so it doesn't get cropped when the OS expects a 16x16 icon. In other words, the tray icon is going to look funky if you're on Linux because Java doesn't support it very well. So with this in mind, I don't think it's even worth it to make a separate tray icon is used in Linux.

**Comment by Hubcapp** [comment]
Alright, here's some before and after for you @Talkarcabbage. This fix could potentially be useful for anyone who doesn't realize how much better size_hints_honor = false is (explained in screenshots, short of it is that windows behave better when you have more control over them) and if they are ALSO set floating mode for the workspace they're on, which usually will mean you only have one window open. i.e. not me or probably anyone.

as of 5997ef6 (patch immediately before)

honor_size_hints = false: https://i.sli.mg/TOLoSE.png
honor_size_hints = true: https://i.sli.mg/nexfO7.png

as of 9907281 (latest patch)
honor_size_hints = false: https://i.sli.mg/IRLziJ.png
honor_size_hints = true: https://i.sli.mg/75ECMN.png

What actually needs to happen to fix notifications on a tiling window manager is to stop treating notifications like normal windows, and do whatever jdownloader, jitsi, and others do to make the window manager not treat custom notifications as windows (and implement libnotify!!). You don't need to do this if you don't want to, I'll get around to it post merge (if not soon then later).

PS @thelegendofbrian in one of the screenshots (honor_size_hints = false 990781), the icon is missing. This isn't always the case (unrelated to size hints), I'll let you know if I can figure out why it didn't get put in if it happens again, but for now I guess it's fine.

**Comment by Hubcapp** [comment]
it's better because it doesn't disrupt window tiling, but worse because it's just a gray box with nothing in it.

**Comment by Talkarcabbage** [comment]
So it doesn't render properly as a popup window instead of a utility window? Is that consistent?

**Comment by Hubcapp** [comment]
Yeah, the contents of the popup are not rendered, but it was fine as the utility type.

**Comment by Talkarcabbage** [comment]
Can you confirm that happens every time?

**Comment by thelegendofbrian** [comment]
After the popup situation is dealt with sufficiently, I'm comfortable getting these changes merged.

**Comment by Hubcapp** [comment]
yeah it happens on both my computers running AwesomeWM every time it happens.

It is probably to do with https://awesomewm.org/wiki/Problems_with_Java

**Comment by thelegendofbrian** [comment]
> edit5: popup should be spawned on monitor where current window is focused. Always on the same monitor for now.

For the typical user, I feel that the primary monitor is the monitor the user pays the most attention to. If a notification popup appears, I'd want it to be on the main monitor so I don't have to keep looking to my other monitors for notifications, especially if they might alert me of low health.

**Comment by Hubcapp** [comment]
I think the user is looking at the monitor that they are interacting with but okay.

**Comment by thelegendofbrian** [comment]
I suppose we both have a point, it's a matter of if you're AFKing something like combat or actively looking at the client.

**Comment by Talkarcabbage** [comment]
I had Brian revert that change, but it appears to still exist as an issue in debian under specific circumstances, being, the notification is received prior to showing the config gui. Once it's shown, the notif renders normally. This has existed even prior to the "make it look good and move to edt" commit, but at that point it was simply invisible, thus during testing assumed it wasn't showing the window at all, rather than it simply not rendering anything but still showing.

Or the short version, it's a bug that seems to exacerbate onto more window managers when set as a popup, but an issue nonetheless worth looking into.

If anyone has a thought in reviewing the code as to why "The notification window does not repaint properly until the config window has been [set visible] at least one time" by all means let me know.
If possible, I'd like to fix it before merging. If not possible, post-merge fix it at some point. When it's set as a utility and not popup it only seems to affect AWM.

**Comment by OrN** [comment]
Is this ready to be merged yet?

**Comment by Hubcapp** [comment]
I think a summary of above is that this fork of RSC works great for Windows, and "not awful" for macOS (Sierra came out, it's not OSX anymore) and Linux. Buggy behavior regarding the new notifications can be observed on Linux. macOS is already buggy in the main project (regarding resizing) and no one has complained... but notifications seem "OK" there (don't autodismiss in macOS iirc).

I say merge it, even though it's not bug free (only on unix based os's). Fact is that market share between OSs is 90% windows, 9% macOS, 1% Linux last I checked... Additionally, the broken behavior in this new feature is "optional" (not even on by default) and of a non-critical nature (does not cause crashes). Majority of this patch is totally ready and a large improvement.

I promise to come back later and implement libnotify (solving linux bugginess in this patch) but now's not a good time for me. Bug me if I haven't said anything in a month.

**Comment by thelegendofbrian** [comment]
Yeah, everything works except for visual notifications on some Linux distros. However those who have issues can use the notification sound as a workaround until @Hubcapp gets around to adding libnotify. I'd say it's ready to merge.
